File: examples/kotlin/src/test/kotlin/adr/BoundaryTest.kt
package adr

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class BoundaryTest {

    @Test
    fun `commits before it performs, and stamps the actor exactly once`() {
        val order = mutableListOf<String>()
        val bus = object : Bus {
            override fun append(commands: List<Command>, results: List<ToolResult>) {
                order += "append " + commands.joinToString(",") { "${it::class.simpleName}:${it.by}:${it.id}" }
            }
        }
        val sink = object : Sink {
            override fun perform(effect: Effect, mode: PerformMode) {
                order += "perform ${effect::class.simpleName}"
            }
        }

        val b = Boundary(
            Clock.fixed(42L),
            IdSource.sequential("c"),
            bus,
            sink,
            State.initial(listOf(Ticket("4118", "x", Priority.Normal, TicketStatus.Open))),
        )

        b.onStepFinish(FinishedStep(Actor.Agent, listOf(ToolResult.PrioritySet("4118", Priority.High))))

        // commit (the signed command, actor stamped) strictly before perform
        assertEquals(listOf("append SetPriority:Agent:c1", "perform Log"), order)
        assertEquals(Priority.High, b.state.tickets.getValue("4118").priority)
    }

    @Test
    fun `never forges the actor — a human action carries Human through the same path`() {
        val seen = mutableListOf<Command>()
        val bus = object : Bus {
            override fun append(commands: List<Command>, results: List<ToolResult>) {
                seen += commands
            }
        }
        val sink = object : Sink {
            override fun perform(effect: Effect, mode: PerformMode) {}
        }
        val b = Boundary(
            Clock.fixed(1L),
            IdSource.sequential(),
            bus,
            sink,
            State.initial(listOf(Ticket("4118", "x", Priority.Normal, TicketStatus.Open))),
        )

        b.onStepFinish(FinishedStep(Actor.Human, listOf(ToolResult.PrioritySet("4118", Priority.Urgent))))

        val first = seen.first()
        assertTrue(first is Command.SetPriority && first.by == Actor.Human)
    }
}

File: examples/typescript/test/boundary.test.ts
import { describe, it, expect } from "vitest";
import { initialState } from "../src/domain";
import { Boundary, fixedClock, sequentialIds, type Bus, type Sink } from "../src/boundary";
import type { Command } from "../src/command";
import type { Effect } from "../src/effect";
import type { ToolResult } from "../src/tool-result";

describe("the boundary adapter — the one impure seam", () => {
  it("commits before it performs, and stamps the actor exactly once", () => {
    const order: string[] = [];
    const bus: Bus = {
      append: (cmds) => order.push(`append ${cmds.map((c) => `${c.kind}:${c.by}:${c.id}`).join(",")}`),
    };
    const sink: Sink = { perform: (fx) => order.push(`perform ${fx.kind}`) };

    const b = new Boundary(
      fixedClock(42),
      sequentialIds("c"),
      bus,
      sink,
      initialState([{ id: "4118", body: "x", priority: "Normal", status: { kind: "Open" } }]),
    );

    b.onStepFinish({ actor: "Agent", results: [{ kind: "PrioritySet", ticket: "4118", level: "High" }] });

    // commit (the signed command, actor stamped) strictly before perform
    expect(order).toEqual(["append SetPriority:Agent:c1", "perform Log"]);
    expect(b.state.tickets.get("4118")!.priority).toBe("High");
  });

  it("never forges the actor — a human action carries Human through the same path", () => {
    const seen: Command[] = [];
    const bus: Bus = { append: (cmds) => seen.push(...cmds) };
    const sink: Sink = { perform: (_fx: Effect) => {} };
    const b = new Boundary(
      fixedClock(1),
      sequentialIds(),
      bus,
      sink,
      initialState([{ id: "4118", body: "x", priority: "Normal", status: { kind: "Open" } }]),
    );

    const result: ToolResult = { kind: "PrioritySet", ticket: "4118", level: "Urgent" };
    b.onStepFinish({ actor: "Human", results: [result] });

    expect(seen[0]).toMatchObject({ kind: "SetPriority", by: "Human" });
  });
});
